src/typechecker.erl

```diff
@@ -2079,49 +2130,78 @@ do_type_check_expr_in(Env, ResTy, {map, _, Expr, Assocs} = MapUpdate) ->
+    case expect_record_type(ResTy, Name, Env#env.tenv) of
+        {elem_ty, Rec, Cs1} ->
             {VarBindsList, Css}
                 = lists:unzip(
-                    lists:map(fun ({record_field, _, FieldWithAnno, Expr}) ->
-                                      FieldTy = get_rec_field_type(FieldWithAnno, Rec),
-                                      type_check_expr_in(Env, FieldTy, Expr)
-                              end
-                             ,Fields)
-                   ),
+                lists:map(fun ({record_field, _, FieldWithAnno, Expr}) ->
+                    FieldTy = get_rec_field_type(FieldWithAnno, Rec),
+                    type_check_expr_in(Env, FieldTy, Expr)
+                          end
+                    ,Fields)
+            ),
+            RecordTy = type_record(Name, [Type || {_, _, Type} <- Rec]),
             {VarBinds, Cs2} = type_check_expr_in(Env, RecordTy, Exp),
             {union_var_binds([VarBinds|VarBindsList], Env#env.tenv)
-            ,constraints:combine([Cs1, Cs2|Css])}
+                ,constraints:combine([Cs1, Cs2|Css])};
+        {elem_tys, Tyss, Cs} ->
+            case type_check_record_union_in(Env, Tyss, Fields) of
+                none ->
+                    {Ty, _VB, _Cs} = type_check_expr(Env#env{infer = true}, Record),
+                    throw({type_error, Record, Ty, ResTy});
+                {VBs, Css} ->
+                    {union_var_binds(VBs, Env#env.tenv), constraints:combine([Cs|Css])}
+            end;
+        any -> fuck;
+        {type_error, _} ->
+            throw({type_error,
+                   Record,
+                   type(record, [{atom, erl_anno:new(0), Name}]),
+                   ResTy})
     end;
 do_type_check_expr_in(Env, ResTy, {record_field, Anno, Expr, Name, FieldWithAnno} = RecordField) ->
-    Rec = get_record_fields(Name, Anno, Env#env.tenv),
-    FieldTy = get_rec_field_type(FieldWithAnno, Rec),
-    case subtype(FieldTy, ResTy, Env#env.tenv) of
+    RecFields = get_record_fields(Name, Anno, Env#env.tenv),
+    {FieldIndex, FieldTy} = get_rec_field_index_and_type(FieldWithAnno, RecFields, 2),
+    {RecTy, RefinedFieldTy} =
+        case Expr of
+            {var, P, Var} ->
+                case Env#env.venv of
+                    #{Var := VarType} ->
+                        case VarType of
+                            {type, _, record, [{atom, _, Name}]} ->
+                                {VarType, FieldTy};
+                            {type, _, record, [{atom, _, Name}|RefinedFields]} ->
+                                {VarType, lists:nth(FieldIndex-1, RefinedFields)};
+                            _ ->
+                                throw({type_error, mismatch, type_record(Name), Expr})
+                        end;
+                    _ ->
+                        throw({unknown_variable, P, Var})
+                end;
+            _ -> {type_record(Name), FieldTy}
+        end,
```

Comment:
This is so ugly @_@

Reply:
Yeah, it'd be good to generalize this.

One way to perhaps solve this would be to change `type_check_expr_in` so that it could also return a type, a type that was a refinement, or a subtype, of the type that was passed as an argument. That way we could call `type_check_expr_in` on the expression and get back the refined type of the record and return the appropriate field.

I'm not asking you to fix this now, that's quite a big change. I'm just thinking out loud.

Reply:
> One way to perhaps solve this would be to change type_check_expr_in so that it could also return a type

If we do that, can we also unify `type_check_expr` with `type_check_expr_in`? (with `any()` as the expected type)

Reply:
Ha, yes of course! That's a really good point @zuiderkwast ! I think that would be a very worthwhile rewrite!

So one invariant would be that the type that is returned is always a subtype of the type passed in from the context. There's an interesting corner case wrt `any()`: if the context demands a concrete type that is not `any()` but we infer the type `any()` what should we return? It should always be safe to return the concrete type, I think.

It's a bit daunting to think about rewriting the typechecker but I think it'll be great to do it. I'm not going to jump on it right away because I'm working on the constraint solver right now (branch `constraint_solver` for anyone interested).  So if anyone wants to take this on right now, that'd be really cool.

Reply:
If we keep the current behaviour, is should be fine, right? We have a lot of test cases so I think attempting this kind of refactoring is pretty safe.

Reply:
Yes, I agree. We have a pretty good test suite at this point.

Reply:
Seems I found a test case that require implementing this. As soon as you have two records embedded, this fails because it only checks for a variable. This examples cannot pass now.

```erlang
-record(r2, {value :: undefined | integer()}).
-record(r1, {r2 :: undefined | #r2{}}).

-spec test(#r1{}) -> integer().
test(#r1{r2 = undefined}) -> 
    0;
test(#r1{r2 = #r2{value = undefined}}) -> 
    0;
% Works
% test(#r1{value = #r2{value = Value}}) ->
%     Value.
test(R1) ->
    R1#r1.r2#r2.value.
```

Reply:
Alright I think I fixed this. I went the `type_check_expr` route instead of `type_check_expr_in`. Eveything passes and the new test case passes. It also fixed a case in our codebase 👍

Reply:
If you guys are fine with it, I would resolve this comment. The idea of refactoring for something cleaner could be tackled later.